### DISCUSSION TITLE
How can I Get system resource(CPU/Memory/Temperature/..) from ios 17

### DISCUSSION THREAD

**Discussion by HaoWShi** [doronz88/pymobiledevice3#621] [Q&A]
I want to get system resource when I execute my test cases, How can I get it .
Very much looking forward to a reply

**Comment by HaoWShi** [comment]
I find something, but I do not know whether it is right
I execute command :  python -m pymobiledevice3 developer dvt device-information
but  I get the following error message

Could not find platform dependent libraries <exec_prefix>
 pymobiledevice3.cli.remote[12432] WARNING create_core_device_tunnel_service failed to be imported. Some feature may not work.
You can debug this by trying the import yourself:

from pymobiledevice3.remote.core_device_tunnel_service import create_core_device_tunnel_service
__main__[12432] ERROR Failed to access an invalid lockdown service, possibly from DeveloperDiskImage.dmg or a Cryptex. You may try: python3 -m pymobiledevice3 mounter auto-mount

**Reply by CKylinMC** [reply]
I'm looking for a solution for some similar use case. I tried this command with `--rsd` option on MacOS and successed (iPad mini 6 with latest iOS 17), it output a JSON:

```
{
    "hardware": {
        "hwCPU64BitCapable": 1,
        "hwCPUsubtype": 2,
        "hwCPUtype": 16777228,
        "numberOfCpus": 6,
        "numberOfPhysicalCpus": 6
    },
    "kernel-name": "/mach.release.t8110",
    "kpep-database": null,
    "network": {
        "en0": "Wi-Fi",
        "en1": "Ethernet Adapter (en1)",
        "en2": "Ethernet Adapter (en2)",
        "lo0": "Loopback"
    }
}
```

I'm not seeing any information relatived to CPU usages or memory usages here. Maybe some other command could do this?

**Reply by HaoWShi** [reply]
yes，As for battery temperature， there is a solution
`python3 -m pymobiledevice3 diagnostics battery monitor`
 and you can get info by 1hz
for example:
-- {'InstantAmperagr':-141, 'Temperature': 3980, 'Voltage': 4447, 'IsCharging': False, CurrentCapacity': 100}
And  the unit of temperature is in Celsius. 3980 for example means 39.8℃

**Comment by doronz88** [comment]
I believe you're addressing one of the following commands:

```shell
pymobiledevice3 developer dvt sysmon process single
pymobiledevice3 developer dvt energy <PID LIST>
```

Please note that starting at iOS 17, all the developer commands now require an RSD tunnel as specified in:
https://github.com/doronz88/pymobiledevice3#working-with-developer-tools-ios--170

**Reply by HaoWShi** [reply]
OK, But I read the code and But I find that all of the commands you mentioned require process information and then returning the resource information of the corresponding process. 
In fact, I also want to get resource usage information for the entire device.

**Reply by doronz88** [reply]
The first one gives you a process snapshot. You can then query these pids for energy

**Comment by HaoWShi** [comment]
As for battery temperature，there is a solution

```shell
python3 -m pymobiledevice3 diagnostics battery monitor
```

and you can get info by 1hz
for example:

```python
{'InstantAmperagr':-141, 'Temperature': 3980, 'Voltage': 4447, 'IsCharging': False, CurrentCapacity': 100}
```

And the unit of temperature is in Celsius. 3980 for example means 39.8℃

As for CPU usage, I need I tunnel connect first. 

By referring to this method(https://github.com/doronz88/pymobiledevice3/blob/master/misc/RemoteXPC.md#reusing-the-macos-trusted-tunnel), you can reuse existing. if not , you need to create tunnel by yourself with command:

```shell
sudo python3 -m pymobiledevice3 remote start-quic-tunnel
```
You can get info as following:

- RSD Address: fd7b:e5b:6f53::1
- RSD Port: 64337

And then execute the command:

```shell
python3 -m pymobiledevice3 developer dvt sysmon process monitor --rsd fd7b:e5b:6f53::1 64337 0.01
```

Parameter Description:
- `fd7b:e5b:6f53::1` means the RSD Address
- `64337` means the RSD Port
- `0.01` means the threshold, and information about processes whose CPU usage exceeds the threshold is printed.

For example:

```python
[process(pid=1771, name='AutomationModeUI', cpuUsage=0.032), pid=91, name='bluetoothd', cpuUsage=1.98)]
```
